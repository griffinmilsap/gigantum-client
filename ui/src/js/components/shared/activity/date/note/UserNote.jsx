// @flow
// vendor
import React, { Component } from 'react';
import { WithContext as ReactTags } from 'react-tag-input';
import classNames from 'classnames';
// mutations
import CreateUserNoteMutation from 'Mutations/activity/CreateUserNoteMutation';
import { setErrorMessage } from 'JS/redux/actions/footer';
// assets
import './UserNote.scss';

let simple;

type Props = {
  changeFullScreenState: Function,
  datasetId: string,
  editorFullscreen: Boolean,
  labbookId: string,
  name: string,
  owner: string,
  sectionType: string,
  toggleUserNote: Function,
}

class UserNote extends Component<Props> {
  state = {
    tags: [],
    userSummaryText: '',
    addNoteDisabled: true,
    unsubmittedTag: '',
  };

  /**
    @param {}
    after component mounts apply simplemde to the dom element id:markdown
  */
  componentDidMount() {
    const { changeFullScreenState } = this.props;

    import('simplemde').then((comp) => {
      if (document.getElementById('markDown')) {
        const Simple = comp.default;
        simple = new Simple({
          element: document.getElementById('markDown'),
          spellChecker: true,
        });
        const fullscreenButton = document.getElementsByClassName('fa-arrows-alt')[0];
        const sideBySideButton = document.getElementsByClassName('fa-columns')[0];

        if (fullscreenButton) {
          fullscreenButton.addEventListener('click', () => changeFullScreenState());
        }

        if (sideBySideButton) {
          sideBySideButton.addEventListener('click', () => changeFullScreenState(true));
        }
      }
    });
  }

  /**
    @param {}
    calls CreateUserNoteMutation adds note to activity feed
  */
  _addNote = () => {
    const { state } = this;
    const self = this;
    const tags = state.tags.map(tag => (tag.text));
    const {
      name,
      owner,
      sectionType,
      labbookId,
      datasetId,
      toggleUserNote,
    } = this.props;
    const {
      unsubmittedTag,
      userSummaryText,
    } = this.state;

    if (unsubmittedTag.length) {
      this._handleAddition({
        id: unsubmittedTag,
        text: unsubmittedTag,
      });
    }


    this.setState({
      addNoteDisabled: true,
      unsubmittedTag: '',
    });

    CreateUserNoteMutation(
      sectionType,
      name,
      userSummaryText,
      simple.value(),
      owner,
      [],
      tags,
      labbookId || datasetId,
      (response, error) => {
        toggleUserNote(false);
        self.setState({
          tags: [],
          userSummaryText: '',
          addNoteDisabled: false,
        });
        if (error) {
          setErrorMessage(owner, name, 'Unable to unlink dataset', error);
        }
      },
    );
  }

  /**
    @param {number} index
    removes tag from list
  */
   _handleDelete = (index) => {
     const { tags } = this.state;

     tags.splice(index, 1);

     this.setState({ tags });
   }

   /**
     @param {String} tag
     add tag to list
   */
   _handleAddition = (tag) => {
     const { tags } = this.state;
     tags.push(tag);
     this.setState({ tags, unsubmittedTag: '' });
   }

   /**
     @param {String} tag
     @param {Number} curPos
     @param {Number} newPos
     drags tag to new position.
   */
   _handleDrag = (tag, currPos, newPos) => {
     const { tags } = this.state;

     // mutate array
     tags.splice(currPos, 1);
     tags.splice(newPos, 0, tag);

     // re-render
     this.setState({ tags });
   }


   /**
     @param {String} unsubmittedTag
     sets unsubmittedTag to state
   */
   _updateUnsubmittedTag = (unsubmittedTag) => {
     this.setState({ unsubmittedTag });
   }

   /**
     @param {Object} event
     calls updates state for summary text
     and enables addNote button if > 0
   */
   _setUserSummaryText = (evt) => {
     const summaryText = evt.target.value;
     this.setState({
       userSummaryText: summaryText,
       addNoteDisabled: (summaryText.length === 0),
     });
   }


   render() {
     const { toggleUserNote, editorFullscreen } = this.props;
     const { addNoteDisabled, tags } = this.state;
     // declare css here
     const userNoteSummaryCSS = classNames({
       UserNote__summary: true,
       'UserNote__summary--fullscreen': editorFullscreen,
     });
     return (
       <div className="UserNote flex flex--column">

         <div
           role="presentation"
           className="UserNote__close close"
           onClick={() => toggleUserNote(false)}
         />

         <input
           type="text"
           placeholder="Add a summary title"
           onKeyUp={evt => this._setUserSummaryText(evt)}
           className="UserNote__title"
         />
         <div className={userNoteSummaryCSS}>
           <textarea
             ref={(ref) => { this.markdown = ref; }}
             id="markDown"
           />
         </div>

         <ReactTags
           id="TagsInput"
           tags={tags}
           handleInputChange={this._updateUnsubmittedTag}
           handleDelete={(index) => { this._handleDelete(index); }}
           handleAddition={(tag) => { this._handleAddition(tag); }}
           handleDrag={(tag, currPos, newPos) => { this._handleDrag(tag, currPos, newPos); }}
         />
         <div className="BtnContainer">
           <button
             type="submit"
             className="Btn Btn--flat"
             onClick={() => toggleUserNote(false)}
           >
             Cancel
           </button>
           <button
             type="submit"
             className="Btn Btn--last"
             disabled={addNoteDisabled}
             onClick={() => { this._addNote(); }}
           >
             Add Note
           </button>
         </div>
       </div>
     );
   }
}

export default UserNote;
