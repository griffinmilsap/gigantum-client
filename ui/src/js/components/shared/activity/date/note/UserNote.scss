@import '~Styles/_imports.scss';

.UserNote {
  padding: 10px 0 0 0;
  position: relative;
  flex: 1;
  width: calc(100% - 40px);

  animation: slide-down-fade-in 0.25s;
  .fa {
    font-family: FontAwesome !important;
  }

  &__summary--fullscreen {
    position: fixed !important;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;

    z-index: index($elements, modal);

    height: 100vh;
    width: 100vw;
  }
}
.UserNote {
  .UserNote__title[type="text"] {
    color: $primaryText;
    margin: 10px 0 20px;
    width: 100% !important;
  }

}


.UserNote__close {
  position: absolute;
  top: -10px !important;
  right: -10px !important;
  cursor: pointer;
  z-index: index($elements, button);
  height: 24px;
  width: 24px;

  @include close();

}

.UserNote__container {
  padding: 18px 0;
  position: relative;
}

.CodeMirror, .CodeMirror-scroll {
  min-height: 250px !important;
}

.CodeMirror-sizer {
  height: 200px;
}

.UserNote {
  span.ReactTags__tag:first-child{
      margin-left: 0px !important;
  }

  div.ReactTags__selected a.ReactTags__remove {
      color: #aaa;
      margin-left: 5px;
      cursor: pointer;
  }

  div.ReactTags__suggestions {
      position: absolute;
  }
  div.ReactTags__suggestions ul {
      list-style-type: none;
      box-shadow: .05em .01em .5em rgba(0,0,0,.2);
      background: white;
      width: 200px;
  }
  div.ReactTags__suggestions li {
      border-bottom: 1px solid #ddd;
      padding: 5px 10px;
      margin: 0;
  }
  div.ReactTags__suggestions li mark {
      text-decoration: underline;
      background: none;
      font-weight: 600;
  }
  div.ReactTags__suggestions ul li.ReactTags__activeSuggestion {
      background: #b7cfe0;
      cursor: pointer;
  }

  .ReactTags__selected {
      display: flex;
      height: auto;
      top: 100px;
      border: 1px solid $grey;
      border-top: 1px solid $primary;
      padding: 0 10px;
      overflow: auto;
    }

  .ReactTags__tag.Languages,
  .ReactTags__tag.Dataset.Type {
      background: $primary !important;
      color: $white !important;
      a {
          color: $white !important;
      }
  }

  .ReactTags__tag.Development.Environments {
      background: $slateBlue !important;
      color: $white !important;
      a {
          color: $white !important;
      }
  }

  .ReactTags__tag.Tags {
      background: $lightblue !important;
      color: $white !important;
      a {
          color: $white !important;
      }
  }
}
