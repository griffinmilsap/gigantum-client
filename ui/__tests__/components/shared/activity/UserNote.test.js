// vendor
import React from 'react'
import renderer from 'react-test-renderer';
import { mount } from 'enzyme'
import relayTestingUtils from '@gigantum/relay-testing-utils'
// components
import UserNote from 'Components/shared/activity/date/note/UserNote';


describe('UserNote', () => {
  it('Renders a snapshot', () => {
    const wrapper = renderer.create(<UserNote />);

    const tree = wrapper.toJSON();

    expect(tree).toMatchSnapshot();
  });

});
