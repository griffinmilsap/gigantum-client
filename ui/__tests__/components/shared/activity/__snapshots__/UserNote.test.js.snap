// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserNote Renders a snapshot 1`] = `
<div
  className="UserNote flex flex--column"
  data-selenium-id="UserNotex"
>
  <div
    className="UserNote__close close"
    onClick={[Function]}
    role="presentation"
  />
  <input
    className="UserNote__title"
    onKeyUp={[Function]}
    placeholder="Add a summary title"
    type="text"
  />
  <div
    className="UserNote__summary"
  >
    <textarea
      id="markDown"
    />
  </div>
  <div
    className="ReactTags__tags react-tags-wrapper"
  >
    <div
      className="ReactTags__selected"
    >
      <div
        className="ReactTags__tagInput"
      >
        <input
          aria-label="Add new tag"
          className="ReactTags__tagInputField"
          id="TagsInput"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          placeholder="Add new tag"
          type="text"
        />
      </div>
    </div>
  </div>
  <div
    className="BtnContainer"
  >
    <button
      className="Btn Btn--flat"
      onClick={[Function]}
      type="submit"
    >
      Cancel
    </button>
    <button
      className="Btn Btn--last"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Add Note
    </button>
  </div>
</div>
`;
